### DISCUSSION TITLE
Extremely Long CFAST Runs

### DISCUSSION THREAD

**Discussion by Sari-Alkhatib** [firemodels/cfast#2118] [General]
I would like to check the opinions on when CFAST runs take extremely long time, i.e., extremely small time steps in the magnitude of 10^(-7) to 10^(-9).

I am encountering such cases in two situations:
- When there is limited ventilation or flow to the compartment and the fire is large, i.e., under-ventilated fires. I observed that with small natural vents defined in the lower portion of the compartment, CFAST would be unable to complete the run (I left it for one whole day and it did not move 1 sec in the simulation). I think this issue is similar to what was discussed [here ](https://groups.google.com/g/cfast/c/be9hQPZKxio/m/mgeT7yM4AAAJ) and [here](https://groups.google.com/g/cfast/c/be9hQPZKxio/m/mgeT7yM4AAAJ). 
- When the fire base is above the layer height for long time during the fire time (i.e., very large flame length ratio). I am not sure if the fire location or the flame impingement are the main issues here. I still think it is about the ventilation and oxygen levels in the upper zone (same as the first point above) since if there is good ventilation, these cases would run just fine.

Are there any documentations or stated limitations/guidelines that would help understanding this behavior in CFAST in both cases?

Thanks in advance!

**Comment by bewygs** [comment]
I'm also experiencing the same issue where CFAST runs take extremely long times with very small time steps for under-ventilated cases (example: a simulation that run for a entire month and never end [slow.txt](https://github.com/user-attachments/files/15934466/slow.txt))

Currently, it seems there is no available solution to speed up the process other than changing some input values to be outside of under-ventilated cases. 🙁

Previously, I was using version 7.1.2 of CFAST and did not encounter this issue. Perhaps running your specific case with this version might help ?

However, the problem with that old version is that I encountered another error that caused the CFAST simulation to stop for some random inputs values. This issue has since been resolved as of #669 .

I personally tried modifying the source code to manually fix the time step dt, so it does not go below a threshold value. Unfortunately, I encountered the same issue as with CFAST version 7.1.2.

From what I understand, the simulation runs through a while loop until the final time you indicate is reached.
https://github.com/firemodels/cfast/blob/e50b50279eb1a7580138cba0be8b02c1486c958e/Source/CFAST/solve.f90#L366

I might be wrong, but inside this while loop, various checks and computation are made. One important check determines which is the value of the next time step to compute inside `ddassl` subroutine.

https://github.com/firemodels/cfast/blob/e50b50279eb1a7580138cba0be8b02c1486c958e/Source/CFAST/numerics.f90#L44

The error I encountered when putting a treshold to the next time step to compute is this one : https://github.com/firemodels/cfast/blob/e50b50279eb1a7580138cba0be8b02c1486c958e/Source/CFAST/numerics.f90#L611-L613

I don't know if there is a way to bypass this verification while not violating the model output.

I aslo use gprof to see which routine takes the most time to runs and I got this for a long run (that I killed before it finished).

```
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls  Ts/call  Ts/call  name    
 71.96    210.06   210.06                             hflow_routines_mp_wall_flow_
  5.58    226.34    16.28                             conduction_routines_mp_conductive_flux_
  2.69    234.19     7.85                             __libm_pow_l9
  1.74    239.28     5.09                             conduction_routines_mp_conduction_
  1.50    243.66     4.38                             mflow_routines_mp_mechanical_flow_
  1.14    246.99     3.33                             solve_routines_mp_calculate_residuals_
  0.86    249.50     2.51                             solve_routines_mp_update_data_
  0.84    251.96     2.46                             numerics_routines_mp_dnrm2_
  0.79    254.26     2.30                             fire_routines_mp_do_fire_
  0.74    256.41     2.15                             hflow_routines_mp_flogo_
  0.73    258.54     2.13                             fire_routines_mp_fire_
  0.70    260.59     2.05                             numerics_routines_mp_ddot_
....
```

**Comment by preneke** [comment]
I would be interested in seeing the gprof analysis for a number of runs. Ones that run fast and ones your are having problems with. There are several issues to be aware or when you do this. One is the amount of time the code spends calculating any particular specific phenomena is directly impacted by the scenario specified. Your analysis suggests that the case you ran has a lot of doors and windows, and possible leaks. Because CFAST is solving an ODE my expectation is that the ratio of different phenomena called will be related to the case. I would be interested to see my hypothesis proved. 

Our experience is that oxygen limiting of the fire is the biggest culprit of slow runs. Combustion is a very locally specific problem and it is difficult for a model using broad generalizations to solve. We are working on this but I don't want to make any promises or implications. At the moment be aware it is a known limitation and we have not given up on dealing with it.  

Also, I would suggest that you not try to set the DT. The solver is making the DT that small because it cannot converge to an answer with a larger DT. It increases the DT when it can. There MIGHT be times when the DT can or should be larger than the solver uses but if those situations exist they are going to be for very specific conditions and not a general case.

**Reply by Sari-Alkhatib** [reply]
Regarding the lower oxygen limit, I am using a value of 1% (assuming it allow more oxygen for the fire and therefore can be considered conservative). I guess I can test changing it to the default value of 15%?

Additionally, in my scenario I have only one natural vent connecting to the outside, and another one that connects to another compartment. So, I am not sure if the number of vents is the issue in my case, but I definitely change their dimensions, areas, and locations in my sensitivity analysis.

**Reply by preneke** [reply]
1% is beyond conservative. In terms of running time I have no idea if 15% or 10% will change things. It might make it slower, might make it faster. 

When we talk about vents we mean all vents. So if the other room is the only other room with a single other vent to the outside and you don't have leaks set you would have three vents. If you have a lot of other rooms with lots of vents and leaks in all rooms you would have a lot of vents. 

In terms of the time, I am saying that the ratio of time spent for each phenomena is related to the number of cases of the phenomena, for example vents, the scenario has. I should have been clear that I don't know the ratio each phenomena has in terms of computational time. I am kind of assuming it is all roughly the same. If you don't have a lot of vents the timing would suggest that the slowest part of CFAST calculations for each time step is the vents. Interesting but not super important since most cases run fairly fast and it isn't the vent calculations that are leading to your slow case.

**Reply by Sari-Alkhatib** [reply]
I see. In my case, I only have one natural vent connecting to the 'outside' and another vent that connects the two compartments I have. However, their areas and elevations varies. 

So far, the cases that CFAST struggles with are definitely connected to extremely under-ventilation conditions, but they also have the upper layer going under the fire base for long times. (At least this is my observation so far.)

**Reply by preneke** [reply]
Yes, when the upper lay gets down to or below the bottom of the fire problems are created. There are a number of situations, that rarely happen, where the simulation hits a discontinuity. An easy one to understand is the top of a vent and the upper layer. If the fire is is burning just hot enough that it gets the upper layer to the top of the vent but when the upper layer drops below the top of the vent the flow out the vent from the upper layer is greater than the flow into the upper layer from the fire the layer will rise above the top of the vent. If the fire and vent size are perfectly matched and the fire is not growing what happens is the solver takes very small steps and the upper layer drops below the top of the vent in one step and rises above the top of the vent in the next step. The same kind of thing is happening with the fires in you long running cases. There is just enough change in each step to either increase or decrease the amount of fuel that burns and the amount of fuel that burns changes the conditions to the opposite happens the next step. It is more convoluted but that is essentially what is happening. In you case, if you are not concerned with door jet fires to the outside you could reduce the HRR in the input to more closely match the amount of burning that CFAST is giving you. I think that it will speed up the run. The amount of unburned fuel in the layer will be wrong but if that is not what you are worried about it won't be a problem.

**Reply by Sari-Alkhatib** [reply]
I think this is exactly what is going on in some of my runs. I am not concerned with door jet fires, but also, in the context my sensitivity analysis, I can't just reduce the HRR. I decided to isolate these cases and discuss them as "cases that challenges CFAST solver," would that make sense?

If such conditions/configurations (that challenge the solver) are obtained using a real-life case (e.g., based on a precise data collection of a certain compartment), and basically, you have limited freedom on what to modify and adjust in the input structure, would you recommend investigating it further using CFAST? Or a more advanced tool would be needed?

**Reply by preneke** [reply]
That is perfectly reasonable. My suggest just to be clear and not to push to do it, would be to find the point where it slows down and adjust the HRR there, meaning set a timepoint where you want to start lowering the HRR and set the HRR at what it would be originally then set another point further on that is say 110% of what the actual HRR is. work that way. It could take some time but it might give you a better idea of what the actual hazard might be. Obviously it is your call and just discussing the cases CFAST is struggling with is a completely reasonable thing to do.

**Reply by Sari-Alkhatib** [reply]
Thanks a lot! I think my problem is now clear. 

Just one more question, what "key performance measures" would be recommend to judge this issue? For example, one thing I am doing is to compare the 'expected' HRR I inputted into CFAST vs the 'actual' HRR and evaluate the difference between them (given that I want my fire to be fully developed). Another example, based on our discussion here, I am now extracting the relative relationship between the layer height and 1) the fire base height, and 2) the upper edge of the single natural vent (both of which are random in my study). Would there be any other parameters that can strongly connect the influential and relevant factors in this case?

**Reply by preneke** [reply]
The only other thing I would add is to look at the oxygen level in the layers, both upper and lower. Otherwise I think you are covering it.

**Reply by Sari-Alkhatib** [reply]
I guess that makes sense too. Thanks a lot. This was extremely helpful!
